src/gordon_gcp/clients/http.py

```diff
@@ -95,15 +92,96 @@ def __init__(self, auth_client=None, session=None):
             body (obj): (optional) A dictionary, bytes, or file-like
                 object to send in the body of the request.
             headers (dict): (optional) HTTP headers to send with the
-                request. Headers pass through to the request will
+                request. Headers passed through to the request will
                 include :attr:`DEFAULT_REQUEST_HEADERS`.
+            retry_type (str): (optional) The type of retries to use: "none",
+                "simple", (retry up to tries/timeout), "no4xx" (don't retry
+                on 4xx, otherwise up to tries/timeout), or "custom".  If
+                "custom", uses the retry predicate specified in the other
+                parameters.  See also extra_401_retry.
+            tries (int): (optional) Total maximum number of times to try the
+                request, including the first (i.e. 1 == no retries).  The number
+                of actual tries is subject to retry_type and timeout as well as
+                this value.
+            timeout (int): (optional) Total maximum seconds to keep trying the
+                request (None == unlimited).  The number of actual tries is
+                subject to retry_type and tries as well as this value.
+            retry_predicate (callable): (optional) Function to call to decide
+                if we should keep trying the request.  Ignored unless
+                retry_type is "custom".  Must return True for "continue retrying
+                (subject to tries and timeout)" or False for "stop retrying".
+                Last argument must be the response from _request().  If the
+                function takes any other arguments, their values must be
+                supplied in predicate_args and/or predicate_kwargs.
+            predicate_args (list): (optional) See retry_predicate.
+            predicate_kwargs (dict): (optional) See retry_predicate.
+            extra_401_retry (bool): (optional) If True, run one more request
+                attempt on a 401 (unauthorized), regardless of the value of
+                retry_type.  This makes sure we've tried generating a new token.
```

Comment:
If we can't make decorators work at all, then I suggest lumping all `backoff`-related keyword arguments in one, something like `backoff_kwargs={...}`. Then creating a helper method to parse all that is needed. I completely understand the preference to be explicit, but this method is a bit overloaded and deceptively complex, and the keyword args defined are not necessarily obviously applicable to backoff functionality. Having it all in one kwarg would remove the helpful docstrings, so for that, I'd suggest writing [prose documentation](https://github.com/spotify/gordon-gcp/blob/develop/docs/clients.rst#asynchronous-gcp-http-client) and referring to that in this docstring.

Reply:
I think this idea is pretty good. I know some other libraries that do a lot of args/kwargs (mock, async-ssh, paramiko) and using them even for simple cases has me looking up the docs almost every time.

Reply:
Makes sense I guess, although I need to figure out what that looks like.  Also not sure I got the message of your prose documentation link?